src/jemalloc.c

```diff
@@ -1706,7 +1714,27 @@ imalloc_body(static_opts_t *sopts, dynamic_opts_t *dopts) {
 	 * some reason.  Let's grab it right away.
 	 */
 	tsd = tsd_fetch();
-	witness_assert_lockless(tsd_tsdn(tsd));
+
+	/*
+	 * If we need to handle reentrancy, we can do it out of a
+	 * known-initialized arena (i.e. arena 0).
+	 */
+	reentrancy_level = tsd_reentrancy_levelp_get(tsd);
+	++*reentrancy_level;
+	if (*reentrancy_level == 1) {
+		witness_assert_lockless(tsd_tsdn(tsd));
+	}
+	if (unlikely(*reentrancy_level > 1)) {
+		/*
+		 * We should never specify particular arenas or tcaches from
+		 * within our internal allocations.
+		 */
+		assert(dopts->tcache_ind == TCACHE_IND_AUTOMATIC);
+		assert(dopts->arena_ind = ARENA_IND_AUTOMATIC);
+		dopts->tcache_ind = TCACHE_IND_NONE;
```

Comment:
It helps minimizes the number of things that we have to reason about (albeit in a vague, sort of ill-defined way).

For example: suppose we detect that the tcache is empty. We go to the arena, and become reentrant. We reentrantly do some operations that fill the tcache. We then finish being reentrant, and return to the initial allocation, having filled the tcache. Now, the reason the initial allocation went into the arena in the first place is because its tcache was empty -- I think it's pretty reasonable for us to assume at that point that the tcache is still empty.